sshd-core/src/main/java/org/apache/sshd/common/forward/SocksProxy.java

```diff
@@ -278,23 +286,25 @@ protected void onMessage(Buffer buffer) throws IOException {
                 if (debugEnabled) {
                     log.debug("Received socks5 connection message");
                 }
-                super.onMessage(buffer);
+                
+                synchronized (this) {
+                    if (delayedBuffers != null) {
+                        if (debugEnabled) {
+                            log.debug("Delaying socks5 connection message");
+                        }
+                        delayedBuffers.add(buffer);
+                    } else {
+                        super.onMessage(buffer);
+                    }
+                }
             }
         }
-
-        @SuppressWarnings("synthetic-access")
-        protected void onChannelOpened(OpenFuture future) {
+        
+        private void sendChannelSuccess() {
             int wpos = response.wpos();
             response.rpos(0);
             response.wpos(1);
-            Throwable t = future.getException();
-            if (t != null) {
-                service.unregisterChannel(channel);
-                channel.close(true);
-                response.putByte((byte) 0x01);
-            } else {
-                response.putByte((byte) 0x00);
-            }
+            response.putByte((byte) 0x00);
```

Comment:
Seems that channel success is sent back to the peer **before** `onChannelOpened` is invoked - i.e., if the channel open fails, the peer is simply brutally disconnected after (erroneously) thinking the the channel has been opened. I am not familiar with SOCKS protocol, but can the peer send messages before receiving an OK that channel has been opened.

Reply:
Not sure if this was written before or after my other reply, pasting in here for clarity.

HostA (SOCKS) -> SSHD -> HostB

HostA requests connection via SOCKS/SSHD to Host B - SocksInitRequest
SSHD replies SocksInitReply
HostA requests command connect via SOCKS/SSHD
SSHD connects to HostB
HostB sends first packets to SSHD 
SSHD sends these packets to HostA
SSHD replies SOCKS command connection open to HostA

As I said the patch could be confusing because I don't fix the underlying issue.  I work around it using a different flow.

 

The underlying issue is that the packets from HostB are sent to HostA, before SHHD sends the socks command response to HostA.

So the better fix would be to delay the packets from HostB until SSH has sent the socks command response to HostA.

However I couldn't see an easy way to do this.

 

So instead of this in the patch I immediately send the socks command response to HostA, to make sure its always sent before any packets from HostB.

The issue is now HostA thinks that the socks connection is active when actually we could be still opening the connection to HostB.

So HostA could start sending packets to HostB at this point.

So I then have to buffer these packets until the connection to HostB is actually open, at which point i then send on the packets.


> Seems that channel success is sent back to the peer before onChannelOpened is invoked

Yes I did this to ensure that the the channel success message is always received by the client before any packets from the channel.

> if the channel open fails, the peer is simply brutally disconnected after (erroneously) thinking the the channel has been opened. 

yes, as I said in the original comment "Note: as we are just using SSHD for testing I wasn't overly concerned with error checking or avoiding synchronization - I'm just providing it to help illustrate the issue"

> I am not familiar with SOCKS protocol, but can the peer send messages before receiving an OK that channel has been opened.

No it can't.  But that isn't the issue.
The issues is that the server is sending packets from the channel before the channel OK message is sent.